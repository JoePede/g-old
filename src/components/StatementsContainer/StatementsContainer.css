.list {
  > div:not(:last-child) {
    margin-bottom: 2rem;
  }
}
