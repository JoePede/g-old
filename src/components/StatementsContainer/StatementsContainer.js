import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Statement from '../Statement';
import {
  deleteStatement,
  flag,
  solveFlag,
  createStatement,
  updateStatement,
} from '../../actions/statement';
import { createLike, deleteLike } from '../../actions/statement_like';
import { updateUser } from '../../actions/user';
import { getStatementUpdates, getVisibibleStatementsByPoll } from '../../reducers';

class StatementsContainer extends React.Component {
  static propTypes = {
    statements: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        text: PropTypes.string,
        pollId: PropTypes.string,
      }),
    ),
    poll: PropTypes.shape({
      likedStatements: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.string,
          statementId: PropTypes.string,
        }),
      ),
      ownStatement: PropTypes.shape({ id: PropTypes.string }),
      ownVote: PropTypes.shape({
        id: PropTypes.string,
      }),
    }).isRequired,

    user: PropTypes.shape({ id: PropTypes.string }).isRequired,

    followees: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      }),
    ).isRequired,
    updates: PropTypes.shape({}).isRequired,
    hideOwnStatement: PropTypes.bool,
    updateStatement: PropTypes.func.isRequired,
    createStatement: PropTypes.func.isRequired,
    deleteStatement: PropTypes.func.isRequired,
    solveFlag: PropTypes.func.isRequired,
    flag: PropTypes.func.isRequired,
    createLike: PropTypes.func.isRequired,
    deleteLike: PropTypes.func.isRequired,
    updateUser: PropTypes.func.isRequired,
  };

  static defaultProps = {
    statements: [],
    ownVote: null,
    ownStatement: null,
    likedStatements: [],
    hideOwnStatement: false,
  };

  constructor(props) {
    super(props);
    this.state = {};
    this.handleStatementSubmission = this.handleStatementSubmission.bind(this);
  }
  handleStatementSubmission({ id, text }, update) {
    if (update) {
      this.props.updateStatement({ id, text });
    }
  }
  render() {
    const {
      poll: { ownStatement, ownVote, id, likedStatements, mode },
      user,
      hideOwnStatement,
      followees,
      updates,
      statements,
    } = this.props;
    let ownStatementsNode = null;
    let statementsList = [];

    if (mode.withStatements) {
      if (!ownStatement && ownVote) {
        const data = {
          vote: ownVote,
          author: user,
          pollId: id,
          likes: 0,
          text: '',
        };
        ownStatementsNode = (
          <Statement
            user={user}
            key={'0000'}
            updates={updates['0000']}
            {...data}
            onCreate={this.props.createStatement}
            asInput
            followees={followees}
          />
        );
      }
      if (ownStatement && !hideOwnStatement) {
        ownStatementsNode = (
          <Statement
            user={user}
            onSubmit={this.handleStatementSubmission}
            key={ownStatement.id}
            {...ownStatement}
            updates={updates[ownStatement.id]}
            ownStatement
            followees={followees}
            onUpdate={this.props.updateStatement}
            onCreate={this.props.createStatement}
            onDelete={this.props.deleteStatement}
          />
        );
      }
      // eslint-disable-next-line eqeqeq
      statementsList = statements.filter(s => s.author.id != user.id);
    }
    return (
      <div>
        {ownStatementsNode}
        {statementsList.map(s =>
          (<Statement
            {...s}
            user={user}
            key={s.id}
            ownLike={likedStatements && likedStatements.find(data => data.statementId === s.id)}
            onLike={this.props.createLike}
            onDeleteLike={this.props.deleteLike}
            onFollow={this.props.updateUser}
            onDelete={this.props.deleteStatement}
            onModeration={this.props.solveFlag}
            onFlagging={this.props.flag}
            followees={followees}
            isFollowee={followees.some(f => f.id === s.author.id)}
          />),
        )}
      </div>
    );
  }
}

const mapStateToProps = (state, { poll: { id }, filter }) => {
  const statements = getVisibibleStatementsByPoll(state, id, filter);
  if (statements) {
    statements.sort((a, b) => b.likes - a.likes);
  }
  return {
    statements,
    updates: getStatementUpdates(state),
  };
};
const mapDispatch = {
  deleteStatement,
  flag,
  solveFlag,
  createStatement,
  updateStatement,
  createLike,
  deleteLike,
  updateUser,
};

export default connect(mapStateToProps, mapDispatch)(StatementsContainer);
