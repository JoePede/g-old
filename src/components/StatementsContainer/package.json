{
  "name": "StatementsContainer",
  "version": "0.0.0",
  "private": true,
  "main": "./StatementsContainer.js"
}
