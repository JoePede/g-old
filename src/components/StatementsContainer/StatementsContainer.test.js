/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */
import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IntlProvider } from 'react-intl';

import StatementsContainer from './StatementsContainer';
import App from '../../components/App';

const middlewares = [thunk.withExtraArgument({ fetch: () => {} })];
const mockStore = configureStore(middlewares);
const setupApp = (store, element) => (
  <IntlProvider locale="de-DE" initialNow={new Date()}>
    <App
      context={{
        intl: {
          initialNow: 1501773564065,
          formattedMessage: () => 'translated text',
          formatDate: () => 'formatted date',
          formatTime: () => 'formatted time',
          formatRelative: () => 'formatted relative',
          formatNumber: () => 'formatted number',
          formatPlural: () => 'formatted plural',
          formatMessage: () => 'formatted message',
          formatHTMLMessage: () => 'formatted html',
          now: () => new Date(),
          locale: 'de-DE',
          newLocale: null,
          messages: {
            'de-DE': {},
          },
        },
        fetch: () => {},
        insertCss: () => {},
        store,
      }}
    >
      {element}
    </App>
  </IntlProvider>
);

const defaultUser = {
  id: 1,
  role: {
    type: 'user',
  },
};

const defaultPollingMode = {
  id: 1,
  withStatements: true,
};
const defaultPoll = {
  id: 1,
  upvotes: 0,
  downvotes: 0,
  likedStatements: null,
  ownStatement: null,
  ownVote: null,
  closedAt: null,
  mode: '1',
};

const initialState = {
  runtime: {
    availableLocales: ['de-DE'],
  },
  intl: {
    locale: 'de-DE',
  },
  entities: {
    roles: {
      2: {
        type: 'user',
      },
    },
    users: {
      byId: {
        1: {
          id: '1',
          name: 'firstOne',
          surname: 'lastOne',
          thumbnail: 'https://api.adorable.io/avatars/32/firstOne.io.png',
          role: 3,
        },
        2: {
          id: '2',
          name: 'firstTwo',
          surname: 'lastTwo',
          thumbnail: 'https://api.adorable.io/avatars/32/firstTwo.io.png',
          role: 3,
        },
      },
    },
    votes: {
      1: {
        id: '1',
        position: 'con',
      },
      2: {
        id: '2',
        position: 'con',
      },
    },
    polls: {
      1: {
        ...defaultPoll,
      },
    },
    statements: {
      byId: {
        1: {
          __typename: 'StatementDL',
          id: '1',
          likes: 5,
          text: 'before',
          pollId: '1',
          createdAt: 'Sat Jul 29 2017 23:52:50 GMT+0200 (CEST)',
          updatedAt: null,
          vote: '1',
          author: '1',
        },
        2: {
          __typename: 'StatementDL',
          id: '2',
          likes: 4,
          text: 'before',
          pollId: '1',
          createdAt: 'Sat Jul 29 2017 23:52:50 GMT+0200 (CEST)',
          updatedAt: null,
          vote: '2',
          author: '2',
        },
      },
      byPoll: {
        1: {
          all: [1, 2],
        },
      },
    },

    proposals: {
      byId: {
        1: { __typename: 'ProposalDL' },
      },
    },
    pollingModes: {
      1: { ...defaultPollingMode },
    },
    activities: {
      allIds: { ids: [1] },
      byId: {
        1: {
          id: '1',
          type: 'statement',
          objectId: '1',
          verb: 'create',
          createdAt: 'Sat Jun 17 2017 09:09:50 GMT+0200 (CEST)',
          actor: '1',
          object: {
            id: '1',
            schema: 'StatementDL',
          },
        },
      },
    },
  },
  ui: {},
};

const initPoll = {
  ...defaultPoll,
  mode: {
    ...defaultPollingMode,
  },
};

const defaultFollowees = [];

describe('StatementsContainer', () => {
  it('Displays statements', () => {
    const store = mockStore(initialState);
    const wrapper = renderer
      .create(
        setupApp(
          store,
          <StatementsContainer
            poll={initPoll}
            user={defaultUser}
            filter={'all'}
            followees={defaultFollowees}
          />,
        ),
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
  it('Merges new Statements from feed correctly', () => {
    let store = mockStore(initialState);
    let wrapper = renderer
      .create(
        setupApp(
          store,
          <StatementsContainer
            poll={initPoll}
            user={defaultUser}
            filter={'all'}
            followees={defaultFollowees}
          />,
        ),
      )
      .toJSON();
    let asString = JSON.stringify(wrapper);
    expect(asString).toMatch(/before/);

    // change statement
    const updatedState = {
      ...initialState,
      entities: {
        ...initialState.entities,
        statements: {
          ...initialState.entities.statements,
          byId: {
            ...initialState.entities.statements.byId,
            3: {
              __typename: 'StatementDL',
              id: '3',
              likes: 0,
              text: 'after',
              pollId: '1',
              createdAt: 'Sat Jul 29 2017 23:52:50 GMT+0200 (CEST)',
              updatedAt: null,
              vote: '2',
              author: '2',
            },
          },
          byPoll: {
            1: {
              all: [3, 2, 1],
            },
          },
        },
      },
    };
    store = mockStore(updatedState);
    wrapper = renderer
      .create(
        setupApp(
          store,
          <StatementsContainer
            poll={initPoll}
            user={defaultUser}
            filter={'all'}
            followees={defaultFollowees}
          />,
        ),
      )
      .toJSON();
    asString = JSON.stringify(wrapper);
    expect(asString).not.toMatch(/before/);
    expect(asString).toMatch(/after/);
  });
});
