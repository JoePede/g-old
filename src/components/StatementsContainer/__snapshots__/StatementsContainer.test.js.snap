// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatementsContainer Displays statements 1`] = `
<div>
  <div
    className="rootAlt contra"
  >
    <div
      onClick={[Function]}
      role="button"
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <img
        alt="IMG"
        className="avatar"
        src="https://api.adorable.io/avatars/32/firstTwo.io.png"
      />
    </div>
    <div
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="header"
      >
        <div>
          <span
            className="author"
          >
            firstTwo
             
            lastTwo
          </span>
          <span>
             (+4)
          </span>
        </div>
        <div
          className="root column"
          onClick={null}
        >
          <button
            className="button plain"
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="icon"
            >
              <svg
                height="24px"
                role="img"
                viewBox="0 0 24 24"
                width="24px"
              >
                <path
                  d="M3,13 L3,11 L5,11 L5,13 L3,13 Z M11,12.9995001 L11,11 L12.9995001,11 L12.9995001,12.9995001 L11,12.9995001 Z M19,12.9995001 L19,11 L20.9995001,11 L20.9995001,12.9995001 L19,12.9995001 Z"
                  fill="none"
                  stroke="#666"
                  strokeWidth="2"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        className="text"
      >
        <div
          className="expanded"
        >
          before
        </div>
      </div>
    </div>
  </div>
</div>
`;
